File: Week 6/pset6/credit.py
#!/usr/bin/env python3
# ----------------------------------------------------------
# |     University of Amsterdam: Minor Programmeren        |
# |     Problem Set 6: Credit (more)                       |
# |     Created by Davy Boekhout on 11/10/2017             |
# |                                                        |
# |     Checks if a given credit number is valid and is    |
# |     either from American Express, Mastercard or Visa.  |
# ----------------------------------------------------------


def get_cc_number():
    """Returns the credit card number the user entered."""
    while True:
        number = input("Number: ")
        if number.isdigit():
            return number


def main():
    """Applies Luhn's algorithm to check a credit card number and prints
    the vendor to stdout.
    """
    # define valid prefixes for credit cards
    AMEX = [34, 37]
    MCRD = [50, 51, 52, 53, 54, 55]
    VISA = [4]

    # set credit card properties
    cc_number = get_cc_number()
    cc_prefix = int(str(cc_number[:2]))
    cc_length = len(cc_number)
    cc_string = ""
    cc_summed = 0

    # apply Luhn's check algorithm by multiplying every other digit
    cc_part_1 = [str(int(x) * 2) for x in cc_number[-2::-2]]

    # and then saving the remaining numbers we haven't processed yet
    if cc_length % 2 == 0:
        cc_part_2 = [x for x in cc_number[1::2]]
    else:
        cc_part_2 = [x for x in cc_number[::2]]

    # sum results in cc_part_1 and cc_part_2
    for digit in cc_part_1:
        cc_string += digit
    for digit in cc_string:
        cc_summed += int(digit)
    for digit in cc_part_2:
        cc_summed += int(digit)

    # if cc_summed ends with 0, high probability of valid cc_number
    if cc_summed % 10 == 0:
        if cc_prefix in AMEX and cc_length == 15:
            print("AMEX")
        elif cc_prefix in MCRD and cc_length == 16:
            print("MASTERCARD")
        elif int(cc_prefix / 10) in VISA and \
                (cc_length == 13 or cc_length == 16):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == '__main__':
    main()

File: Week 1/pset1/credit.c
/*
----------------------------------------------------------
|     University of Amsterdam: Minor Programmeren        |
|     Problem Set 1: Credit                              |
|     Created by Davy Boekhout on 5/09/2017              |
|                                                        |
|     Checks a given credit card number validity         |
|     and identifies the corresponding credit            |
|     card company.                                      |
----------------------------------------------------------
*/

#include <stdio.h>
#include <cs50.h>

int main (void)
{
    // Get credit card number
    printf("Number: ");
    long long input = get_long_long();
    long long credit_card = input;
    long long credit_card2 = input / 10;
    long long credit_card3 = input;
    int number_of_digits = 0;
    int total_value = 0;

    // Determine how long the number is
    while (input !=0)
    {
        input /= 10;
        number_of_digits++;
    }

    // Add the numbers, always starting with the last one
    int add = 0;
    while (credit_card != 0)
    {
        add = add + (credit_card % 10);
        credit_card = credit_card / 100;
    }

    // Multiply the numbers, starting second-to-last
    int multiply = 0;
    int multiply_total = 0;
    while (credit_card2 != 0)
    {
        multiply = (credit_card2 % 10) * 2;
        if (multiply >= 10)
            multiply_total = multiply_total + multiply - 9;
        else
            multiply_total += multiply;
        credit_card2 = credit_card2 / 100;
    }

    // Getting the sum of multiplying and adding according to Hans Peter Luhn algorithm
    total_value += add + multiply_total;

    // Determine if credit card number is valid and what company it is from (AMEX/MASTERCARD/VISA)
    if (total_value % 10 == 0)
    {
        int count = number_of_digits;
        // Looping once more over the credit card number to get the first two digits
        while (count > 0)
        {
            credit_card3 = credit_card3 / 10;
            count--;
            if (count == 2)
            {
                int first_two = credit_card3;
                int first_digit = credit_card3 / 10;
                // Switching to determine credit card company
                switch (first_two)
                {
                    case 34:
                    case 37:
                        printf("AMEX\n");
                        break;
                    case 51:
                    case 52:
                    case 53:
                    case 54:
                    case 55:
                        printf("MASTERCARD\n");
                        break;
                    default:
                        if (first_digit == 4)
                            printf("VISA\n");
                        else
                            printf("INVALID\n");
                }
            }
        }
    }
    else
    {
        printf("INVALID\n");
    }
}